src/server/tiered_storage.cc

```diff
@@ -454,55 +519,61 @@ void TieredStorage::WriteSingle(DbIndex db_index, PrimeIterator it, size_t blob_
     return;
   }
 
-  constexpr size_t kMask = kBlockAlignment - 1;
-  size_t page_size = (blob_len + kMask) & (~kMask);
+  SingleRequest* req = new SingleRequest(blob_len, res, it->first.ToString());
 
-  DCHECK_GE(page_size, blob_len);
-  DCHECK_EQ(0u, page_size % kBlockAlignment);
+  auto& enqueued_entries = db_arr_[db_index]->bigbin_enqueued_entries;
+  auto emplace_res = enqueued_entries.emplace(req->key, req);
+  CHECK(emplace_res.second);
 
-  struct SingleRequest {
-    char* block_ptr = nullptr;
-    PrimeTable* pt = nullptr;
-    size_t blob_len = 0;
-    off_t offset = 0;
-    string key;
-  } req;
+  it->second.GetString(req->block_ptr);
+  it->second.SetIoPending(true);
 
-  char* block_ptr = (char*)mi_malloc_aligned(page_size, kBlockAlignment);
+  auto cb = [this, req, db_index](int io_res) {
+    PrimeTable* pt = db_slice_.GetTables(db_index).first;
 
-  req.blob_len = blob_len;
-  req.offset = res;
-  req.key = it->first.ToString();
-  req.pt = db_slice_.GetTables(db_index).first;
-  req.block_ptr = block_ptr;
+    absl::Cleanup cleanup = [this, req]() {
```

Comment:
it looks like a custom deleter for req, maybe it will be better to use unique_ptr with a custom deleter?